clang/test/Sema/code_align.c

```diff
@@ -0,0 +1,164 @@
+  [[clang::code_align(-922337203685477)]]
+  for(int I=0; I<256; ++I) { bar(I); }
+
+  // cpp-local-error@+3{{expression is not an integral constant expression}}
+  // cpp-local-note@+2{{left shift of negative value -1311768467294899680}}
+  // c-local-error@+1{{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was -(__int128_t)1311768467294899680ULL << 64}}
+  [[clang::code_align(-(__int128_t)0x1234567890abcde0ULL << 64)]]
+  for(int I=0; I<256; ++I) { bar(I); }
+
+  // cpp-local-error@+3{{expression is not an integral constant expression}}
+  // cpp-local-note@+2{{function parameter 'A' with unknown value cannot be used in a constant expression}}
+  // c-local-error@+1{{expression is not an integer constant expression}}
+  [[clang::code_align(A)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+}
+
+void check_code_align_expression() {
+  int a[10];
+
+  // Test that checks expression is not a constant expression.
+  int foo2; // cpp-local-note {{declared here}}
+  // c-local-error@+3{{expression is not an integer constant expression}}
+  // cpp-local-error@+2{{expression is not an integral constant expression}}
+  // cpp-local-note@+1{{read of non-const variable 'foo2' is not allowed in a constant expression}}
+  [[clang::code_align(foo2 + 1)]]
+  for (int i = 0; i != 10; ++i)
+    a[i] = 0;
+
+#if __cplusplus >= 201103L
+  // Test that checks expression is a constant expression.
+  constexpr int bars = 0;
+  [[clang::code_align(bars + 1)]]
+  for (int i = 0; i != 10; ++i)
+    a[i] = 0;
+#endif
+}
+
+#if __cplusplus >= 201103L
+template <int A, int B, int C, int D, int E>
+void code_align_dependent() {
+  [[clang::code_align(C)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  [[clang::code_align(A)]] // OK
+  [[clang::code_align(A)]] // OK
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  [[clang::code_align(A)]]
+  [[clang::code_align(E)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // cpp-local-error@+1{{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was 23}}
+  [[clang::code_align(B)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+
+  // cpp-local-error@+2{{'code_align' attribute requires an integer argument which is a constant power of two between 1 and 4096 inclusive; provided argument was -10}}
+  // cpp-local-note@#neg-instantiation{{in instantiation of function template specialization}}
+  [[clang::code_align(D)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+}
+
+template<int ITMPL>
+void bar3() {
+  [[clang::code_align(8)]]
+  [[clang::code_align(ITMPL)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+}
+
+template<int ITMPL1>
+void bar4() {
+  [[clang::code_align(ITMPL1)]]
+  [[clang::code_align(32)]]
+  for(int I=0; I<128; ++I) { bar(I); }
+}
+
+int main() {
+  code_align_dependent<8, 23, 32, -10, 64>(); // #neg-instantiation
+  bar3 <4>();
```

Comment:
Why didn't this and the one below it cause a diagnostic?  This results in the two having different alignments, so shouldn't the duplicate check have caught it during instantiation?

Reply:
Yes, it is missing diagnostic!

Reply:
@erichkeane, i have fixed this diagnostic issues with commit https://github.com/llvm/llvm-project/pull/70762/commits/6778154ef0d64052f62b75955a615d542bafb799